<a href="https://colab.research.google.com/github/xslittlemaggie/Deep-Learning-Projects/blob/master/DNN_Mnist.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Step 1: Import librarious

In [0]:
import numpy as np
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.layers import Conv2D, MaxPooling2D, Flatten
from keras.optimizers import SGD, Adam
from keras.utils import np_utils
from keras.datasets import mnist

## Step 2: Load data

In [0]:
def load_data():
  (x_train, y_train), (x_test, y_test) = mnist.load_data()
  number = 10000
  x_train = x_train[0:number]
  y_train = y_train[0:number]
  x_train = x_train.reshape(number, 28 * 28)
  x_test = x_test.reshape(x_test.shape[0], 28 * 28)
  x_train = x_train.astype('float32')
  x_test = x_test.astype('float32')
  # convert class vectors to binary class matrices
  y_train = np_utils.to_categorical(y_train, 10)
  y_test = np_utils.to_categorical(y_test, 10)
  
  x_train = x_train
  x_test = x_test
  
  x_train = x_train/255   # normalize the input, if not normalized, the accuracy decreases significantly
  x_test = x_test/255
  #x_test = np.random.normal(x_test)  # add random noise to testing data, this would decrease the test accuracy
  return (x_train, y_train), (x_test, y_test)

In [0]:
(x_train, y_train), (x_test, y_test) = load_data()

## Step 3: Get familiar with the data

In [0]:
print("x_train shape:", x_train.shape)
print("y_train shape:", y_train.shape)
print("x_test shape:", x_test.shape)
print("y_test shape:", y_test.shape)

x_train shape: (10000, 784)
y_train shape: (10000, 10)
x_test shape: (10000, 784)
y_test shape: (10000, 10)


In [0]:
print(y_train[0]) # the first value is 5

[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]


## Step 4: Create the DNN model

Models:
      **1). Dense -> fully connected NN**
      **2). MaxPooling**
      **3). Conv2D -> CNN**
      **4). Flatten**


1. Activation functions: 
      **1). relu** (more efficient)
      **2). sigmoid**
      **3). tanh**
      **4). softmax** (usually the last layer)
      
2. Loss functions: 
      **1). mse** (not good for classification)
      **2). categorical_crossentropy**
      
3. Optimizers: 
      **1). SGD(lr = 0.01)**
      **2). Adam**
      
  

In [0]:
model = Sequential()
model.add(Dense(input_dim = 28 * 28, units = 512, activation = 'relu'))  # only need to add the input_dim at the first layer)
#model.add(Dropout(0.2))  # if there is overfitting, use dropout, add dropout at every hidden

model.add(Dense(units = 512, activation = 'relu'))
#model.add(Dropout(0.2))

model.add(Dense(units = 512, activation = 'relu'))
#model.add(Dropout(0.2))

#for i in range(10):  # add more layers, not work
#  model.add(Dense(units = 512, activation = 'relu'))   


model.add(Dense(units = 10, activation = 'softmax')) # the units (neurons) of the last layer need to be the # of classes

model.compile(loss = 'categorical_crossentropy', optimizer = "adam", metrics = ['accuracy'])

model.fit(x_train, y_train, batch_size = 100, epochs = 30)

result = model.evaluate(x_train, y_train, batch_size = 10000)
print("Train Acc:", result[1])
print()

result = model.evaluate(x_test, y_test, batch_size = 10000)
print("Test Acc:", result[1])

Epoch 1/30
10000/10000 [==============================] - 1s 102us/step - loss: 0.4608 - acc: 0.8656
Epoch 2/30
10000/10000 [==============================] - 1s 70us/step - loss: 0.1524 - acc: 0.9561
Epoch 3/30
10000/10000 [==============================] - 1s 69us/step - loss: 0.0958 - acc: 0.9700
Epoch 4/30
10000/10000 [==============================] - 1s 71us/step - loss: 0.0536 - acc: 0.9843
Epoch 5/30
10000/10000 [==============================] - 1s 68us/step - loss: 0.0319 - acc: 0.9903
Epoch 6/30
10000/10000 [==============================] - 1s 70us/step - loss: 0.0407 - acc: 0.9868
Epoch 7/30
10000/10000 [==============================] - 1s 68us/step - loss: 0.0213 - acc: 0.9936
Epoch 8/30
10000/10000 [==============================] - 1s 69us/step - loss: 0.0182 - acc: 0.9938
Epoch 9/30
10000/10000 [==============================] - 1s 70us/step - loss: 0.0228 - acc: 0.9931
Epoch 10/30
10000/10000 [==============================] - 1s 71us/step - loss: 0.0201 - acc: 0.993

Result analysis:

**Train** acc: 1

**Test** acc: 0.97

It looks good!

